session.go

```diff
@@ -0,0 +1,219 @@
+package connmux
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+)
+
+// session encapsulates the multiplexing of streams onto a single "physical"
+// net.Conn.
+type session struct {
+	net.Conn
+	windowSize  int
+	pool        BufferPool
+	out         chan []byte
+	streams     map[uint32]*stream
+	closed      map[uint32]bool
+	connCh      chan net.Conn
+	beforeClose func(*session)
+	mx          sync.RWMutex
+}
+
+// startSession starts a session on the given net.Conn using the given transmit
+// windowSize and pool. If connCh is provided, the session will notify of new
+// streams as they are opened. If beforeClose is provided, the session will use
+// it to notify when it's about to close.
+func startSession(conn net.Conn, windowSize int, pool BufferPool, connCh chan net.Conn, beforeClose func(*session)) *session {
+	s := &session{
+		Conn:        conn,
+		windowSize:  windowSize,
+		pool:        pool,
+		out:         make(chan []byte),
+		streams:     make(map[uint32]*stream),
+		closed:      make(map[uint32]bool),
+		connCh:      connCh,
+		beforeClose: beforeClose,
+	}
+	go s.sendLoop()
+	go s.recvLoop()
+	return s
+}
+
+func (s *session) recvLoop() {
+	for {
+		b := s.pool.getForFrame()
+		// First read id
+		id := b[:idLen]
+		_, err := io.ReadFull(s, id)
+		if err != nil {
+			s.onSessionError(err, nil)
+			return
+		}
+
+		isACK := false
+		isRST := false
+		switch frameType(id) {
+		case frameTypeACK:
+			isACK = true
+		case frameTypeRST:
+			// Closing existing connection
+			isRST = true
+		}
+		setFrameType(id, frameTypeData)
+
+		_id := binaryEncoding.Uint32(id)
+		if isACK {
+			c, open := s.getOrCreateStream(_id)
+			if !open {
+				// Stream was already closed, ignore
+				continue
+			}
+			c.sb.ack <- true
+			continue
+		} else if isRST {
+			s.mx.Lock()
+			c := s.streams[_id]
+			delete(s.streams, _id)
+			s.closed[_id] = true
+			s.mx.Unlock()
+			if c != nil {
+				// Close, but don't send an RST back the other way since the other end is
+				// already closed.
+				c.close(false, nil, nil)
+			}
+			continue
+		}
+
+		// Read frame length
+		dataLength := b[idLen:frameHeaderLen]
+		_, err = io.ReadFull(s, dataLength)
+		if err != nil {
+			s.onSessionError(err, nil)
+			return
+		}
+
+		_dataLength := int(binaryEncoding.Uint16(dataLength))
+
+		// Read frame
+		b = b[:frameHeaderLen+_dataLength]
+		_, err = io.ReadFull(s, b[frameHeaderLen:])
```

Comment:
In theory, `ReadFull` can still read a few bytes even if there's an error from the underlie connection. Maybe still submit the bytes read to the caller if that happens?